src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateResolver.cs

```diff
@@ -485,5 +443,197 @@ private static void AddParameterMatchingToTemplates(IReadOnlyCollection<ITemplat
                 }
             }
         }
+
+        /// <summary>
+        /// The method makes all short names in template group available for matching for given template (using <see cref="TemplateInfoWithGroupShortNames"/> <see cref="ITemplateInfo"/> implementation).
+        /// TODO: double-check if that is needed. If needed that should be replaced by TemplateGroup logic as:
+        /// - group templates by template groups
+        /// - match templates through filters by template group including this logic of sharing short names
+        /// - create resolution result afterwards.
+        /// </summary>
+        /// <param name="templateList">the list of <see cref="ITemplateInfo"/> to process.</param>
+        /// <returns></returns>
+        private static IReadOnlyList<ITemplateInfo> SetupFilterableTemplateInfoFromTemplateInfo(IReadOnlyList<ITemplateInfo> templateList)
+        {
+            Dictionary<string, HashSet<string>> shortNamesByGroup = new Dictionary<string, HashSet<string>>();
+
+            // get the short names lists for the groups
+            foreach (ITemplateInfo template in templateList)
+            {
+                string effectiveGroupIdentity = !string.IsNullOrEmpty(template.GroupIdentity)
+                    ? template.GroupIdentity
+                    : template.Identity;
+
+                if (!shortNamesByGroup.TryGetValue(effectiveGroupIdentity, out HashSet<string> shortNames))
+                {
+                    shortNames = new HashSet<string>();
+                    shortNamesByGroup[effectiveGroupIdentity] = shortNames;
+                }
+
+                if (template is IShortNameList templateWithShortNameList)
+                {
+                    shortNames.UnionWith(templateWithShortNameList.ShortNameList);
+                }
+                else
+                {
+                    shortNames.Add(template.ShortName);
+                }
+            }
+
+            // create the TemplateInfoWithGroupShortNames with the group short names
+            List<TemplateInfoWithGroupShortNames> filterableTemplateList = new List<TemplateInfoWithGroupShortNames>();
+
+            foreach (ITemplateInfo template in templateList)
+            {
+                string effectiveGroupIdentity = !string.IsNullOrEmpty(template.GroupIdentity)
+                    ? template.GroupIdentity
+                    : template.Identity;
+
+                TemplateInfoWithGroupShortNames filterableTemplate = TemplateInfoWithGroupShortNames.FromITemplateInfo(template);
+                filterableTemplate.GroupShortNameList = shortNamesByGroup[effectiveGroupIdentity].ToList();
```

Comment:
If I'm reading this correctly, we are creating a copy of the same list for every template within the same template group here. Can we reuse the same list instead?

Reply:
Yep, we can. We are doing so much refactoring for this class that probably disappear soon - I will be tempted not to delete it when the time comes 😋